### ISSUE TITLE
Support for svt-av1-psy ?

### ISSUE THREAD

**Issue by mr44er** [alexheretic/ab-av1#210]
https://github.com/gianni-rosato/svt-av1-psy
https://wiki.x266.mov/docs/encoders/SVT-AV1-PSY
https://www.reddit.com/r/AV1/comments/1am8f2b/introducing_svtav1psy/

I haven't tried yet, but it should work without modification and it's already in FreeBSD-ports. \o/
What would need modification would likely be the extended/finer CRF steps "Range has been expanded to 70 (from 63) to help with ultra-low bitrate encodes, and can now be set in quarter-step (0.25) increments."

Are the possible steps hardcoded in ab-av1 or is it autodetected?

**Comment by alexheretic** [comment]
> Are the possible steps hardcoded in ab-av1 or is it autodetected?

Can set with `--crf-increment`, similarly `--max-crf` for the other point.

The defaults are hardcoded.

**Comment by mr44er** [comment]
Thx, will check and play with it after all running encodes are done.

Also interesting new things are tune=3 and enable-variance-boost=1, have you experienced how that reacts with VMAF?

**Comment by mr44er** [comment]
```
ab-av1 crf-search --thorough -i t27.mkv 
- crf 32 VMAF 97.47 (17%)
- crf 43 VMAF 95.37 (8%)
- crf 55 VMAF 89.01 (4%)
- crf 44 VMAF 94.93 (8%)
  00:01:17 ################################################################################## (sampling crf 44, eta 0s)
Encode with: ab-av1 encode -i t27.mkv --crf 43

crf 43 VMAF 95.37 predicted video stream size 63.30 MiB (8%) taking 6 minutes

ab-av1 crf-search --thorough -i t27.mkv --crf-increment 0.25
- crf 32.5 VMAF 97.47 (17%)
- crf 43.75 VMAF 94.93 (8%)
- crf 43.5 VMAF 94.93 (8%)
  00:01:17 ############################################################################### (sampling crf 43.25, eta 0s)
Encode with: ab-av1 encode -i t27.mkv --crf 43.25

crf 43.25 VMAF 95.37 predicted video stream size 63.30 MiB (8%) taking 6 minutes
```

Looks like it's working!

**Comment by alexheretic** [comment]
I think that 43 & 43.25 got identical results suggests the codec does not support fractional crf, the encoder/ffmpeg probably just floors the value. If so you'll have faster crf-searches by sticking with the default crf-increment.

**Comment by mr44er** [comment]
Yeah, I saw that and tried multiple files/presets to get any difference shown at all. I'm running some preset0 tests and then I will test if it spits out different file sizes on the fractional stuff.

For now I can report that enable-variance-boost=1 (that is already default with svt-av1-psy) with it's defaults (variance-boost-strength and variance-octile) and tune=3 do a pretty good job.

enable-alt-curve=1 -> bad for VMAF

**Comment by mr44er** [comment]
Ok, fractional crf does not work. Anywhere happens a roundup...just for info, it's not that important for me. ;)
```

ab-av1 encode -i t13.mkv --crf 26 --vfilter "crop=1920:800:0:140" --svt tune=3 --svt enable-overlays=1 --video-only --output crf26.mkv
  00:03:37 ########################################################################################### (28 fps, eta 0s)
Encoded 69.79 MiB (15%)

ab-av1 encode -i t13.mkv --crf 26.25 --vfilter "crop=1920:800:0:140" --svt tune=3 --svt enable-overlays=1 --video-only --output crf26_25.mkv
  00:03:23 ########################################################################################### (30 fps, eta 0s)
Encoded 69.79 MiB (15%)

ab-av1 encode -i t13.mkv --crf 26.75 --vfilter "crop=1920:800:0:140" --svt tune=3 --svt enable-overlays=1 --video-only --output crf26_75.mkv
  00:03:52 ########################################################################################### (27 fps, eta 0s)
Encoded 64.44 MiB (14%)

 ls -l crf*
-rw-r--r--  1 mr44er wheel 73183001 Jun 29 22:55 crf26.mkv
-rw-r--r--  1 mr44er wheel 73183001 Jun 29 23:01 crf26_25.mkv
-rw-r--r--  1 mr44er wheel 67568323 Jun 29 23:05 crf26_75.mkv
```

**Comment by mr44er** [comment]
Maybe some interesting directions to dig deeper. It tells me that variance boost helps to keep more details at small space costs, if VMAF isn't cheated this way and if it's not the regular optimization between the versions.

This search run with svt-av1 2.0.0:

```
ab-av1 crf-search --svt tune=0 --svt enable-overlays=1 --vfilter "crop=1920:816:0:132" -i t00.mkv --preset 0 --thorough
- crf 32 VMAF 94.74 (5%)
- crf 21 VMAF 97.40 (16%)
  04:50:24 ################################################################################## (sampling crf 31, eta 0s)
Encode with: ab-av1 encode -i t00.mkv --crf 31 --preset 0 --vfilter "crop=1920:816:0:132" --svt tune=0 --svt enable-overlays=1

crf 31 VMAF 95.06 predicted video stream size 1.05 GiB (5%) taking 2 days
```

svt-av1-psy 2.1.0:

```
ab-av1 crf-search --svt tune=3 --preset 0 --thorough --svt enable-overlays=1 --vfilter "crop=1920:816:0:132" -i t00.mkv --min-crf 33 
- crf 44 VMAF 91.45 (3%)
- crf 38 VMAF 94.07 (4%)
- crf 33 VMAF 95.53 (6%)
- crf 35 VMAF 94.95 (5%)
  08:02:42 ################################################################################## (sampling crf 34, eta 0s)
Encode with: ab-av1 encode -i t00.mkv --crf 34 --preset 0 --vfilter "crop=1920:816:0:132" --svt tune=3 --svt enable-overlays=1

crf 34 VMAF 95.27 predicted video stream size 1.16 GiB (6%) taking 3 days
```

I won't do both encodings, that would take to long :)

**Comment by stderr-to-devnull** [comment]
@alexheretic returning to the initial request, will you add support for `svt-av1-psy`?

I tried with `ffmpeg` nightly 5.1 but it's not built with `vmaf` so it throws an error. Also tried with `ffmpeg` binaries from `Av1ador` but I also get an error with no `stderr` output.

**Comment by alexheretic** [comment]
> will you add support for svt-av1-psy?

I was under the impression from this thread that if you have svt-av1-psy ffmpeg then ab-av1 is already usable with existing options. What kind of support do you have in mind?

**Comment by stderr-to-devnull** [comment]
>  if you have svt-av1-psy ffmpeg
I don't have it and I don't know how I can obtain it, my technical expertise is limited so I don't know how. My OS is Windows so unless I can find some builds with `vmaf` and `svt-av1-psy` enabled I am at a standstill. What I understand is that this is definitely `ffmpeg` dependent so not something directly related to this project.

In any case, any suggestions on how to proceed is welcome.

Offtopic: regarding GPU accelerated VMAF, I understand it is enabled since `ffmpeg 6.1`. Though maybe in this case, it is more directly related to this project in the sense that a separate `ffmpeg` call should be made available to invoke the GPU accelerated VMAF library instead of the standard software version (if my assumptions are correct on how this is implemented, the information is difficult for me to dig up). I can open a separate issue for this if desired.

**Comment by mr44er** [comment]
> My OS is Windows
Time to peek into other OSes ;)

I don't know how it would work on Windows, I was under the impression it could use the binary .exe (besides ffmpeg).
Community builds are here https://github.com/gianni-rosato/svt-av1-psy/discussions/66

**Comment by mr44er** [comment]
> If you have svt-av1-psy ffmpeg then ab-av1 is already usable with existing options

Yes, the smaller crf-steps don't work, but everything else. Tune3 is king, but that tricks VMAF too low in a false way, so you'll pay with slightly bigger file size. My recommendation is aiming at VMAF93 with tune3 to get a better experience.

**Comment by alexheretic** [comment]
Afaics ab-av1 supports this fairly well with the right ffmpeg, so I'll close this. Let me know if you have suggestions for how ab-av1 could do better here.
